pkg/reconciler/iploop.go

```diff
@@ -101,19 +102,76 @@ func (rl *ReconcileLooper) findOrphanedIPsPerPool(ipPools []storage.IPPool) erro
 func (rl ReconcileLooper) isPodAlive(podRef string, ip string) bool {
 	for livePodRef, livePod := range rl.liveWhereaboutsPods {
 		if podRef == livePodRef {
-			livePodIPs := livePod.ips
-			logging.Debugf(
-				"pod reference %s matches allocation; Allocation IP: %s; PodIPs: %s",
-				livePodRef,
-				ip,
-				livePodIPs)
-			_, isFound := livePodIPs[ip]
-			return isFound || livePod.phase == v1.PodPending
+			isFound := checkIpOnPod(&livePod, podRef, ip)
+
+			if !isFound && (livePod.phase == v1.PodPending) {
+				/* Sometimes pods are still coming up, and may not yet have Multus
+				 * annotation added to it yet. We don't want to check the IPs yet
+				 * so re-fetch the Pod 5x
+				 */
+				var podToMatch *podWrapper = &livePod
+				var retries int = 0
+				logging.Debugf("Re-fetching Pending Pod: %s IP-to-match: %s", livePodRef, ip)
+
+				for retries < storage.PodRefreshRetries {
+					retries += 1
+					podToMatch := rl.refreshPod(livePodRef)
+					if podToMatch == nil {
+						logging.Debugf("Cleaning up...")
+						return false
+					} else if podToMatch.phase != v1.PodPending {
+						logging.Debugf("Pending Pod is now in phase: %s", podToMatch.phase)
+						break
+					} else {
+						isFound = checkIpOnPod(podToMatch, podRef, ip)
+						// Short-circuit - Pending Pod may have IP now
+						if isFound {
+							logging.Debugf("Pod now has IP annotation while in Pending")
+							return true
+						}
+					}
+				}
+				isFound = checkIpOnPod(podToMatch, podRef, ip)
```

Comment:
Would it make sense to use [exponential backoff](https://github.com/kubernetes/apimachinery/blob/ddfe1eccd72c0752f52f4c55845ea9fa73809cca/pkg/util/wait/wait.go#L419) here ? 

Also, I think this could be re-written with k8s primitives:


```golang
import "k8s.io/apimachinery/pkg/util/wait"

timeout      := time.Second
pollInterval := 200*time.Millisecond  // 5 * 0.2 = 1

// PollImmediate tries a condition func until it returns true, an error, or the timeout
// is reached.
// ... 
waitErr := wait.PollImmediate(pollInterval, timeout, func() (bool, error) {
    pod := rl.refreshPod(livePodRef)

    if podToMatch == nil {
        logging.Debugf("Cleaning up...")
        return false, fmt.Errorf("pod is gone")
    } else if isPodPending(pod) {
        logging.Debugf("Pending Pod is now in phase: %s", podToMatch.phase)
        return true, nil
    } else {
        return isIPonPod(pod, podRef, ip), nil // I'm not entirely sure I did not oversimplify this ... 
    }
})
```

Reply:
I am hesitant to use a backoff because remember that for #182 when the IP Pool is full, we WANT it to reconcile. In fact, we dont need any retries in that case... because the container will never come up otherwise. It is in Pending state, does not have an IP annotation either, and will match the podRef. 

So adding more retries or backoff would just add load or delay the reconciler (and thus blocked Pod) longer.

Any suggestions on how to distinguish between these two scenarios? Or should we just retry once